tests/fixtures/app.py

```diff
@@ -0,0 +1,36 @@
+import pytest
+from flask import Flask
+from flask.testing import FlaskClient
+from quizy.app import create_app
+
+
+@pytest.fixture
+def app() -> Flask:
+    app = create_app()
+    app.config.update({
+        "TESTING": True,
+    })
+    yield app
+
+
+@pytest.fixture
+def test_client(app: Flask) -> FlaskClient:
+    return app.test_client()
+
+
+@pytest.fixture
+def test_client_logged_in(test_client: FlaskClient) -> FlaskClient:
```

Comment:
Jeżeli potrzebujesz klienta z zalogowanym użytkownikiem, to prościej byłoby to zrobić w ten sposób: 
https://flask-login.readthedocs.io/en/latest/#automated-testing

Reply:
Tutaj coś nie działa.

Próbuję stworzyć test klienta z zalogowanym użytkownikiem jak w materiałach flaska:

```
import pytest
from flask import Flask
from flask.testing import FlaskClient
from flask_login import FlaskLoginClient
from sqlalchemy.testing.pickleable import User

@pytest.fixture
def logged_in(app: Flask):
    app.test_client_class = FlaskLoginClient
    user = User.query.get(1)
    return app.test_client(user=user)
```

Jednakże, nawet importy od sqlalchemy się nie udają:

```
ImportError while loading conftest '/home/xzdncl/wa/kodilla/kodilla-automated-tests/tests/conftest.py'.
tests/conftest.py:1: in <module>
    from fixtures.app import app, test_client, test_client_logged_in, logged_in
tests/fixtures/app.py:5: in <module>
    from sqlalchemy.testing.pickleable import User
.tox/develop/lib/python3.9/site-packages/sqlalchemy/testing/pickleable.py:12: in <module>
    from . import fixtures
.tox/develop/lib/python3.9/site-packages/sqlalchemy/testing/fixtures.py:29: in <module>
    @config.mark_base_test_class()
.tox/develop/lib/python3.9/site-packages/sqlalchemy/testing/config.py:106: in mark_base_test_class
    return _fixture_functions.mark_base_test_class()
E   AttributeError: 'NoneType' object has no attribute 'mark_base_test_class'
```

sqlalchemy/testing/config.py:23
```
_fixture_functions = None  # installed by plugin_base
```

Niestety jestem zagubiony tutaj.

Spróbowałem tylko lokalnie zaimportować:

```
import pytest
from flask import Flask
from flask.testing import FlaskClient
from flask_login import FlaskLoginClient

@pytest.fixture
def logged_in(app: Flask):
    app.test_client_class = FlaskLoginClient
    from sqlalchemy.testing.pickleable import User
    user = User.query.get(1)
    return app.test_client(user=user)
```

I importy nie przechodzą, podczas tworzenia już fixture dla testu:

```
tests/fixtures/app.py:44: in logged_in
    from sqlalchemy.testing.pickleable import User
.tox/develop/lib/python3.9/site-packages/sqlalchemy/testing/pickleable.py:12: in <module>
    from . import fixtures
.tox/develop/lib/python3.9/site-packages/sqlalchemy/testing/fixtures.py:29: in <module>
    @config.mark_base_test_class()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    def mark_base_test_class():
>       return _fixture_functions.mark_base_test_class()
E       AttributeError: 'NoneType' object has no attribute 'mark_base_test_class'

.tox/develop/lib/python3.9/site-packages/sqlalchemy/testing/config.py:106: AttributeError
```

Reply:
Ja tu widzę kilka problemów, więc ewentualnie możemy spróbować to rozwiązać na rozmowie.

Kilka ścieżek do sprawdzenia na teraz:
1. `from sqlalchemy.testing.pickleable import User` to chyba nie jest właściwy import?
2. Praca z bazą danych w testach [wymaga aktywnego kontekstu aplikacji](https://flask.palletsprojects.com/en/2.1.x/testing/#tests-that-depend-on-an-active-context)
3. `User` musi zostać utworzony i dodany do sesji zanim będziesz mógł wykonać `get`. Sesja niekoniecznie musi zostać skomitowana, ale obiekt musi zostać do niej dodany.

Reply:
Jesteś wstanie rozwinąć kwestię jak to zrobić.
Ponieważ materiały niestety są zbyt ubogie i brakuje im kontekstu.
https://flask-login.readthedocs.io/en/latest/#automated-testing - brakuje tutaj opisu skąd bierze się ten User, nie znalazłem informacji w tym dziale, że musi istnieć w bazie.
Użyłem UserManagera z flask_user, lecz przeglądając teraz dokumentacje do niego nie mam żadnego pojęcia w jaki sposób miałbym z niego skorzystać, aby stworzyć. Użytkownika i następnie go zalogować.

Z mojego punktu widzenia, przerobienie tej części testów nie wpływa na naukę testowania, tylko i wyłącznie na poznanie samej biblioteki flask i jej pochodnych, co nie jest bezpośrednio celem tego szkolenia.
Jeśli zmiana sposobu tworzenia zalogowanego użytkownika nie jest konieczna w dalszej części szkolenia, to w ogóle bym sobie odpuścił przerabianie tego w tym momencie.

Reply:
Chodzi o klasę `User`, którą deklarujesz w `create_app`. W sumie nie ma się za bardzo jak do niej dostać, więc myślę, że sugestia, żeby to na razie zostawić, jest rozsądna.